README.md
# ranged-date

[![NPM](https://nodei.co/npm/ranged-date.png)](https://nodei.co/npm/ranged-date/)

[![NPM](https://img.shields.io/npm/l/ranged-date.svg)](https://www.npmjs.com/package/ranged-date/)
[![NPM](https://img.shields.io/node/v/ranged-date.svg)](https://www.npmjs.com/package/ranged-date/)
[![NPM](https://img.shields.io/npm/dm/ranged-date.svg)](https://www.npmjs.com/package/ranged-date/)
[![NPM](https://img.shields.io/github/last-commit/jpcx/ranged-date.svg)](https://www.npmjs.com/package/ranged-date/)
[![NPM](https://img.shields.io/david/jpcx/ranged-date.svg)](https://www.npmjs.com/package/ranged-date/)
[![NPM](https://img.shields.io/david/dev/jpcx/ranged-date.svg)](https://www.npmjs.com/package/ranged-date/)

**MIGRATION NOTICE:**

-   _ranged-date 2.0.0: back and forward ranges are now mandatory (were 0.5 each by default in 1.x.x)._
-   _ranged-date 2.0.0: Date Objects with values outside of the specified range will return false (would return Date Object in 1.x.x)._

Converts a given string, number, or Date object to the number of milliseconds since the Unix epoch, provided that it can be recognized as millisecond, second, or microsecond time within a specified range from the current date.

Milliseconds, seconds, or microseconds may be excluded as possibilities for range matching.

Useful for:

-   Parsing timestampted data from external APIs.
-   Integrating collections containing mixed timestamp formats.

## Getting Started

### Prerequisites

Node.JS version 6.0.0 or above.

### Installing

    npm i ranged-date

### Testing

The following commands will test the package for errors.

    cd /path/to/node_modules/ranged-date
    npm test

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [ranged-date](#ranged-date)
-   [module:cert-is.TypeAssertionError](#modulecert-istypeassertionerror)
-   [module:cert-is.RangeAssertionError](#modulecert-israngeassertionerror)

### ranged-date

[index.js:150-178](https://github.com/jpcx/ranged-date/blob/2.1.0/index.js#L150-L178 "Source code on GitHub")

Converts a given string, number, or Date object to the number of milliseconds since the Unix epoch, provided that it can be recognized as millisecond, second, or microsecond time within a specified range from the current date.

#### Parameters

-   `data` **([Date](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Date) \| [number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) \| [string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String))** Data to attempt to recognize as valid date.
-   `yrBack` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Years before current date as lower bound.
-   `yrFwd` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Years after current date as upper bound.
-   `exclusions` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** Specifies interpretation exclusions, if any.
    -   `exclusions.us` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Exclude microseconds interpretation. (optional, default `false`)
    -   `exclusions.ms` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Exclude milliseconds interpretation. (optional, default `false`)
    -   `exclusions.s` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Exclude seconds interpretation. (optional, default `false`)

#### Examples

```javascript
// All return current time in ms
rangedDate(new Date(), 1, 1)
rangedDate(Date.now(), 1, 1)
rangedDate(new Date().toString(), 1, 1)
rangedDate(String(Date.now()), 1, 1)

// All return false
rangedDate(Date.now() - 10000000, 0.0000001, 0.0000001)
rangedDate(Date.now(), 1, 1, {ms: true})

// All throw cert-is TypeAssertionErrors
rangedDate()
rangedDate(new Date())
rangedDate({})
rangedDate(328385)

// Throws cert-is RangeAssertionError
rangedDate(Date.now(), -1, -1)
```

-   Throws **([module:cert-is.TypeAssertionError](#modulecert-istypeassertionerror) \| [module:cert-is.RangeAssertionError](#modulecert-israngeassertionerror))** Throws an assertion error if parameter requirements are not met.

Returns **([number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) \| [boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean))** Converted time in ms or false if outside range.

### module:cert-is.TypeAssertionError

[index.js:150-178](https://github.com/jpcx/ranged-date/blob/2.1.0/index.js#L150-L178 "Source code on GitHub")

Thrown by [cert-is](https://github.com/jpcx/cert-is) in response to a type assertion error.

Type: [Error](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Error)

### module:cert-is.RangeAssertionError

[index.js:150-178](https://github.com/jpcx/ranged-date/blob/2.1.0/index.js#L150-L178 "Source code on GitHub")

Thrown by [cert-is](https://github.com/jpcx/cert-is) in response to a range assertion error.

Type: [Error](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Error)

## Versioning

Versioning using [SemVer](http://semver.org/). For available versions, see the [tags on this repository](https://github.com/jpcx/ranged-date/tags).

## Author

-   **Justin Collier** - [jpcx](https://github.com/jpcx)

## License

This project is licensed under the ISC License - see the [LICENSE.md](https://github.com/jpcx/ranged-date/blob/2.0.1/LICENSE.md) file for details


CHANGELOG.md
# Changelog

## <a name="2.1.0"></a> [2.1.0](https://github.com/jpcx/ranged-date/tree/2.1.0) (2019-06-15)

| __[Changes since 2.0.2](https://github.com/jpcx/ranged-date/compare/2.0.2...2.1.0)__ | [Release Notes](https://github.com/jpcx/ranged-date/releases/tag/2.1.0) | [README](https://github.com/jpcx/ranged-date/tree/2.1.0/README.md) |
| --- | --- | --- |

| [Source Code (zip)](https://github.com/jpcx/ranged-date/archive/2.1.0.zip) | [Source Code (tar.gz)](https://github.com/jpcx/ranged-date/archive/2.1.0.tar.gz) |
| --- | --- |

__Features:__

+ __docs:__ Added documentation for cert-is errors.
+ __npm:__  Added documentationjs fix script.
            Added dependency update script.
            Made postinstall migration warning more visible.

## <a name="2.0.2"></a> [2.0.2](https://github.com/jpcx/ranged-date/tree/2.0.2) (2019-06-13)

| __[Changes since 2.0.1](https://github.com/jpcx/ranged-date/compare/2.0.1...2.0.2)__ | [Release Notes](https://github.com/jpcx/ranged-date/releases/tag/2.0.2) | [README](https://github.com/jpcx/ranged-date/tree/2.0.2/README.md) |
| --- | --- | --- |

| [Source Code (zip)](https://github.com/jpcx/ranged-date/archive/2.0.2.zip) | [Source Code (tar.gz)](https://github.com/jpcx/ranged-date/archive/2.0.2.tar.gz) |
| --- | --- |

__Bugfixes:__

+ __docs/npm:__ Fixed documentationjs github hash link generation.

## <a name="2.0.1"></a> [2.0.1](https://github.com/jpcx/ranged-date/tree/2.0.1) (2019-06-13)

| __[Changes since 2.0.0](https://github.com/jpcx/ranged-date/compare/2.0.0...2.0.1)__ | [Release Notes](https://github.com/jpcx/ranged-date/releases/tag/2.0.1) | [README](https://github.com/jpcx/ranged-date/tree/2.0.1/README.md) |
| --- | --- | --- |

| [Source Code (zip)](https://github.com/jpcx/ranged-date/archive/2.0.1.zip) | [Source Code (tar.gz)](https://github.com/jpcx/ranged-date/archive/2.0.1.tar.gz) |
| --- | --- |

__Bugfixes:__

+ __npm:__       Fixed js-yaml dependency versions for documentationjs and eslint.
+ __changelog:__ Fixed bad version in 2.0.0 entry.

## <a name="2.0.0"></a> [2.0.0](https://github.com/jpcx/ranged-date/tree/2.0.0) (2019-06-13)

| __[Changes since 1.1.3](https://github.com/jpcx/ranged-date/compare/1.1.3...2.0.0)__ | [Release Notes](https://github.com/jpcx/ranged-date/releases/tag/2.0.0) | [README](https://github.com/jpcx/ranged-date/tree/2.0.0/README.md) |
| --- | --- | --- |

| [Source Code (zip)](https://github.com/jpcx/ranged-date/archive/2.0.0.zip) | [Source Code (tar.gz)](https://github.com/jpcx/ranged-date/archive/2.0.0.tar.gz) |
| --- | --- |

__Features:__

+ __ranged-date:__ Back and forward ranges are now mandatory (were 0.5 each by default in 1.x.x).
                   Date Objects with values outside of the specified range will return false (would return Date Object in 1.x.x).
+ __docs:__      Moved to documentationjs.
+ __tests:__     Moved to mocha.
+ __changelog:__ Added changelog.


test.js
/**
 * @author Justin Collier <jpcxme@gmail.com>
 * @license ISC
 * @see {@link http://github.com/jpcx/ranged-date|GitHub}
 */

/* .--------------------------------------------------------------------------,
  /                                  Setup                                   /
 '--------------------------------------------------------------------------' */

'use strict'

/* global describe it */

const rangedDate = require('./')
const cert = require('cert-is')

/* .--------------------------------------------------------------------------,
  /                                Test Data                                 /
 '--------------------------------------------------------------------------' */

const testData = (() => {
  const self = {}
  const now = {}
  const later = {}
  now.obj = new Date()
  now.ms = now.obj.valueOf()
  now.s = ~~(now.ms / 1000)
  now.us = now.ms * 1000
  later.obj = new Date(now.ms + 31536000000)
  later.ms = later.obj.valueOf()
  later.s = ~~(later.ms / 1000)
  later.us = later.ms * 1000
  self.obj = {
    name: 'Date Object',
    now: now.obj,
    expected: now.ms,
    later: later.obj
  }
  self.ms = {
    name: 'Date Milliseconds',
    now: now.ms,
    expected: now.ms,
    later: later.ms
  }
  self.s = {
    name: 'Date Seconds',
    now: now.s,
    expected: now.s * 1000,
    later: later.s
  }
  self.us = {
    name: 'Date Microseconds',
    now: now.us,
    expected: now.ms,
    later: later.us
  }
  self.str = {
    name: 'Date Object String',
    now: now.obj.toString(),
    expected: now.s * 1000,
    later: later.obj.toString()
  }
  self.msStr = {
    name: 'Date Milliseconds String',
    now: String(now.ms),
    expected: now.ms,
    later: String(later.ms)
  }
  self.sStr = {
    name: 'Date Seconds String',
    now: String(now.s),
    expected: now.s * 1000,
    later: String(later.s)
  }
  self.usStr = {
    name: 'Date Microseconds String',
    now: String(now.us),
    expected: now.ms,
    later: String(later.us)
  }
  return self
})()

/* .--------------------------------------------------------------------------,
  /                                  Tests                                   /
 '--------------------------------------------------------------------------' */

describe('Range-Match Tests', () => {
  for (let key in testData) {
    const data = testData[key]
    describe(data.name, () => {
      it('should pass adequate ranges', () => {
        cert(rangedDate(data.now, 0.1, 0.1)).is(data.expected)
        cert(rangedDate(data.now, 15, 1)).is(data.expected)
      })
      it('should fail inadequate ranges', () => {
        cert(rangedDate(data.later, 0.5, 0.5)).is(false)
        cert(rangedDate(data.later, 15, 0.5)).is(false)
      })
    })
  }
})

describe('Exclusion Tests', () => {
  for (let key in testData) {
    const data = testData[key]
    describe(data.name, () => {
      const testCase = selfKey => {
        if (!selfKey) {
          // value should always pass
          it('should pass regardless of exclusions', () => {
            cert(
              rangedDate(data.now, 0.1, 0.1, { us: false, ms: false, s: false })
            ).is(data.expected)
            cert(
              rangedDate(data.now, 0.1, 0.1, { us: true, ms: true, s: true })
            ).is(data.expected)
          })
        } else {
          const goodObj = { us: true, ms: true, s: true }
          const badObj = { us: true, ms: true, s: true }
          goodObj[selfKey] = false
          it('should pass when others are excluded', () => {
            cert(rangedDate(data.now, 0.1, 0.1, goodObj)).is(data.expected)
          })
          it('should fail when self is excluded', () => {
            cert(rangedDate(data.now, 0.1, 0.1, badObj)).is(false)
          })
        }
      }
      if (data.name.match(/milliseconds/i)) {
        testCase('ms')
      } else if (data.name.match(/microseconds/i)) {
        testCase('us')
      } else if (data.name.match(/object/i)) {
        testCase(false)
      } else {
        testCase('s')
      }
    })
  }
})


rangedDate.js
'use strict'

/**
 * Returns true only if value falls within range.
 *
 * @param {number} t - Number being tested.
 * @param {number} l - Lower bound.
 * @param {number} u - Upper bound.
 * @example
 * //returns true
 * inRange(0, -1, 1)
 * @example
 * //returns false
 * inRange(-1, 0, 1)
 * @returns {boolean} True if number falls within range, false if not.
 */
const inRange = (t, l, u) => t > l && t < u
/**
 * Returns the difference of a number and the current date in ms, in years from the current date.
 *
 * @param {number} t - Number to subtract from.
 * @example
 * //returns years until unix epoch (negative)
 * yrDist(0)
 * @example
 * //returns 1
 * yrDist(Date.now() + 31536000000)
 * @returns {number} Years until given time in ms.
 */
const yrDist = t => (t - Date.now()) / +'31536e6'
/**
 * Returns the given number only if it falls within a given range in years from the current date.
 *
 * @param {number} t - Number being tested.
 * @param {number} l - Years before current date as lower bound.
 * @param {number} u - Years after current date as upper bound.
 * @example
 * //returns current time in ms
 * bound(Date.now(), 0.5, 0.5)
 * @example
 * //returns false
 * bound(Date.now() * 1000, 0.5, 0.5)
 * @returns {(number|boolean)} Number falls within range.
 */
const bound = (t, l, u) => inRange(yrDist(t), -l, u) && t
/**
 * Returns time in ms if number itself, seconds equivalent, or microseconds equivalent falls within a given range in years from the current date.
 *
 * @param {number} t - Number being tested.
 * @param {number} l - Years before current date as lower bound.
 * @param {number} u - Years after current date as upper bound.
 * @example
 * //returns current time in ms
 * check(Date.now() * 1000, 0.5, 0.5)
 * @example
 * //returns current time in ms
 * check(Date.now() / 1000, 0.5, 0.5)
 * @example
 * //returns false
 * check(Date.now() - 100000000000, 0.5, 0.5)
 * @example
 * //returns current time in ms - 100000000000
 * check(Date.now() - 100000000000, 5, 0.5)
 * @returns {(number|boolean)} Converted ms or false if outside range.
 */
const check = (t, l, u) => isFinite(t) && (
  bound(t, l, u) || bound(t * 1000, l, u) || bound(t / 1000, l, u)
)
/**
 * Converts a given string, number, or date to the number of milliseconds since the Unix epoch, provided that it can be recognized as seconds, milliseconds, or microseconds within a specified range in years from the current date.
 *
 * @exports rangedDate
 * @param {(Date|number|string)} data - Data to attempt to match.
 * @param {number} [low=0.5] - Years before current date as lower bound.
 * @param {number} [up=0.5]  - Years after current date as upper bound.
 * @example
 * //returns current time in ms
 * rangedDate(new Date())
 * @example
 * //returns current time in ms
 * rangedDate(String(new Date()))
 * @example
 * //returns current time in ms
 * rangedDate(Date.now())
 * @example
 * //returns current time in ms
 * rangedDate(String(Date.now()))
 * @example
 * //returns current time in ms
 * rangedDate(Date.now() * 1000)
 * @example
 * //returns current time in ms
 * rangedDate(Date.now() / 1000)
 * @example
 * //returns false
 * rangedDate(Date.now() - 10000, 0.0000001, 0.0000001)
 * @example
 * //returns current time in ms - 10000
 * rangedDate(Date.now() - 10000, 0.000001, 0.000001)
 * @example
 * //returns false
 * rangedDate(10)
 * @example
 * //returns 10
 * const yrSince1970 = new Date().getUTCFullYear() - 1970
 * rangedDate(10, yrSince1970 + 2)
 * @example
 * //returns false
 * rangedDate(Date.now() + 100000000000)
 * @example
 * //returns current time in ms + 100000000000
 * rangedDate(Date.now() + 100000000000, 0.1, 5)
 * @returns {(number|boolean)} Converted time in ms or false if outside range.
 */
module.exports = (data, low = 0.5, up = 0.5) => {
  if (data instanceof Date) return data.valueOf()
  if (typeof data == 'number') return check(data, low, up)
  return check(Date.parse(data), low, up) || check(+data, low, up)
}


scripts/postDocJS.js
/**
 * Formats documentation.js generated Table of Contents for brevity. Fixes source code links and commit-dependent hash links.
 *
 * @private
 * @author  Justin Collier <jpcxist@gmail.com>
 * @license MIT
 * @see {@link http://github.com/jpcx/ranged-date|GitHub}
 */

const fs = require('fs')
const path = require('path')

const pkgVer = require('../package.json').version
const README_PATH = path.join(__dirname, '../README.md')

const README = fs.readFileSync(README_PATH, 'utf8').split('\n')

let section = ''
for (let i = 0; i < README.length; i++) {
  if (README[i].match(/^#+/m)) {
    // Set the section identifier (heading)
    section = README[i].match(/#+ (.*)/m)[1]
  }

  if (section === 'Table of Contents') {
    // Remove Table of Contents Parameters/Properties/Examples
    if (README[i].match(/^ +- +\[(?:Parameters|Properties|Examples)]\(/m)) {
      // Splice README and decrement i
      README.splice(i--, 1)
    }
  } else {
    // Fix source code links
    let sourceCodeMatch
    const updateMatch = () => {
      sourceCodeMatch = README[i].match(
        /^\[.*?.js:(\d+)-(\d+)]\(https:\/\/github.com\/.*?#L(\d+)-L(\d+)/m
      )
    }
    updateMatch()
    if (sourceCodeMatch) {
      // Line is a source code link

      if (sourceCodeMatch[3] === sourceCodeMatch[4]) {
        // Unnecessary from and to line distinction; merge into one line
        README[i] = README[i].replace(
          /^(\[.*?.js:)(\d+)-(\d+)(]\(https:\/\/github.com\/.*?#L)(\d+)-L(\d+)/m,
          '$1$5$4$5'
        )
        updateMatch()
      } else {
        if (sourceCodeMatch[1] !== sourceCodeMatch[3]) {
          // Invalid source code link; match with github
          README[i] = README[i].replace(
            /^(\[.*?.js:)(\d+)/m,
            `$1${sourceCodeMatch[3]}`
          )
          updateMatch()
        }
        if (sourceCodeMatch[2] !== sourceCodeMatch[4]) {
          // Invalid source code link; match with github
          README[i] = README[i].replace(
            /^(\[.*?.js:)(\d+)-(\d+)/m,
            `$1$2-${sourceCodeMatch[4]}`
          )
          updateMatch()
        }
      }
      // fix blob/hash/ -> blob/tag/
      README[i] = README[i].replace(
        /\/blob\/[0-9a-z]+?\//g,
        '/blob/' + pkgVer + '/'
      )
      updateMatch()
    }
  }
}

// Write the new README
fs.writeFileSync(README_PATH, README.join('\n'))


index.js
'use strict'

/**
 * Value certification module. Certifies that parameters passed to ranged-date
 * are of expected types and ranges.
 *
 * @private
 * @module cert-is
 * @external
 * @see {@link https://github.com/jpcx/cert-is}
 */
const cert = require('cert-is')

/**
 * Returns true only if value falls within range.
 *
 * @private
 * @param {number} t - Number being tested.
 * @param {number} l - Lower bound.
 * @param {number} u - Upper bound.
 * @returns {boolean} True if number falls within range, false if not.
 * @example
 * // returns true
 * inRange(0, -1, 1)
 * @example
 * // returns false
 * inRange(-1, 0, 1)
 */
const inRange = (t, l, u) => t > l && t < u
/**
 * Returns the difference of a number and the current date in ms as years from the current date.
 *
 * @private
 * @param {number} target - Number to subtract from.
 * @returns {number} Years until given time in ms.
 * @example
 * // returns years until unix epoch
 * yrDist(0)
 * @example
 * // returns 1
 * yrDist(Date.now() + 31536000000)
 */
const yrDist = target => (target - Date.now()) / +'31536e6'
/**
 * Returns the given number only if it falls within a given range in years from the current date.
 *
 * @private
 * @param {number} target - Number being tested.
 * @param {number} yrBack   - Years before current date as lower bound.
 * @param {number} yrFwd    - Years after current date as upper bound.
 * @returns {(number|boolean)} Number falls within range.
 * @example
 * // returns current time in ms
 * bound(Date.now(), 5, 5)
 * @example
 * // returns false
 * bound(Date.now() * 1000, 5, 5)
 */
const bound = (target, yrBack, yrFwd) =>
  inRange(yrDist(target), -yrBack, yrFwd) && target
/**
 * Returns time in ms if number itself, seconds equivalent, or microseconds equivalent falls within a given range in years from the current date.
 *
 * @private
 * @param   {number}  target     - Number being tested.
 * @param   {number}  yrBack     - Years before current date as lower bound.
 * @param   {number}  yrFwd      - Years after current date as upper bound.
 * @param   {Object}  exclusions - Specifies interpretation exclusions, if any.
 * @param   {boolean} [exclusions.us=false] - Exclude microseconds interpretation.
 * @param   {boolean} [exclusions.ms=false] - Exclude milliseconds interpretation.
 * @param   {boolean} [exclusions.s=false]  - Exclude seconds interpretation.
 * @returns {(number|boolean)}  Converted ms or false if outside range.
 * @example
 * // returns current time in ms
 * check(Date.now(), 5, 5, {})
 * @example
 * // returns current time in ms
 * check(Date.now() / 1000, 5, 5, {})
 * @example
 * // returns current time in ms
 * check(Date.now() * 1000, 5, 5, {})
 * @example
 * // returns false
 * check(Date.now() - 100000000000, 5, 5, {})
 * @example
 * // returns current time in ms - 100000000000
 * check(Date.now() - 100000000000, 5, 5, {})
 * @example
 * // returns false
 * check(Date.now(), 5, 5, { ms: true })
 * @example
 * // returns current time in ms
 * check(Date.now(), 5, 5, { s: true, us: true })
 * @example
 * // returns false
 * check(Date.now() / 1000, 5, 5, { s: true })
 * @example
 * // returns current time in ms
 * check(Date.now() / 1000, 5, 5, { ms: true, us: true })
 * @example
 * // returns false
 * check(Date.now() * 1000, 5, 5, { us: true })
 * @example
 * // returns current time in ms
 * check(Date.now() * 1000, 5, 5, { ms: true, s: true })
 * @example
 * // returns false
 * check(Date.now(), 5, 5, { ms: true, s: true , us: true })
 */
const check = (target, yrBack, yrFwd, exclusions) =>
  isFinite(target) &&
  ((!exclusions.ms && bound(target, yrBack, yrFwd)) ||
    (!exclusions.s && bound(target * 1000, yrBack, yrFwd)) ||
    (!exclusions.us && bound(target / 1000, yrBack, yrFwd)))

/**
 * Converts a given string, number, or Date object to the number of milliseconds since the Unix epoch, provided that it can be recognized as millisecond, second, or microsecond time within a specified range from the current date.
 *
 * @public
 * @module ranged-date
 * @param {(Date|number|string)} data - Data to attempt to recognize as valid date.
 * @param   {number}  yrBack     - Years before current date as lower bound.
 * @param   {number}  yrFwd      - Years after current date as upper bound.
 * @param   {Object}  exclusions - Specifies interpretation exclusions, if any.
 * @param   {boolean} [exclusions.us=false] - Exclude microseconds interpretation.
 * @param   {boolean} [exclusions.ms=false] - Exclude milliseconds interpretation.
 * @param   {boolean} [exclusions.s=false]  - Exclude seconds interpretation.
 * @returns {(number|boolean)} Converted time in ms or false if outside range.
 * @throws  {(module:cert-is.TypeAssertionError|module:cert-is.RangeAssertionError)} Throws an assertion error if parameter requirements are not met.
 * @example
 * // All return current time in ms
 * rangedDate(new Date(), 1, 1)
 * rangedDate(Date.now(), 1, 1)
 * rangedDate(new Date().toString(), 1, 1)
 * rangedDate(String(Date.now()), 1, 1)
 *
 * // All return false
 * rangedDate(Date.now() - 10000000, 0.0000001, 0.0000001)
 * rangedDate(Date.now(), 1, 1, {ms: true})
 *
 * // All throw cert-is TypeAssertionErrors
 * rangedDate()
 * rangedDate(new Date())
 * rangedDate({})
 * rangedDate(328385)
 *
 * // Throws cert-is RangeAssertionError
 * rangedDate(Date.now(), -1, -1)
 */
module.exports = (
  data,
  yrBack,
  yrFwd,
  exclusions = { us: false, ms: false, s: false }
) => {
  cert(data)
    .message('ranged-date data must be a Date, number, or a string')
    .isType(Date, 'number', 'string')
  cert(yrBack, yrFwd)
    .message('ranged-date year boundaries must be specified with numbers > 0')
    .isType('number')
    .isGT(0)
  cert(exclusions)
    .message('ranged-date exclusions must be specified with an object')
    .isType(Object)
  cert(exclusions.s, exclusions.ms, exclusions.us)
    .message('ranged-date exclusions must be booleans')
    .isType('boolean', 'undefined')
  if (data instanceof Date) {
    return check(data.valueOf(), yrBack, yrFwd, {})
  }
  if (typeof data === 'number') return check(data, yrBack, yrFwd, exclusions)
  // data is string
  return (
    check(Date.parse(data), yrBack, yrFwd, {}) ||
    check(+data, yrBack, yrFwd, exclusions)
  )
}

/**
 * Thrown by {@link https://github.com/jpcx/cert-is cert-is} in response to a type assertion error.
 *
 * @public
 * @typedef {Error} module:cert-is.TypeAssertionError
 * @external
 */

/**
 * Thrown by {@link https://github.com/jpcx/cert-is cert-is} in response to a range assertion error.
 *
 * @public
 * @typedef {Error} module:cert-is.RangeAssertionError
 * @external
 */
